src/instrumenter/transpile.ts

```diff
@@ -443,19 +491,128 @@ function transpileLet(expr: SLet, ctx: TranspilingContext): Expression {
+
+    const body: Block = factory.makeBlock([]);
+
+    // Build and insert a for-loop with empty body
+    return factory.makeForStatement(body, initStmt, iterCond, incStmt);
+}
+
+function makeMemberAccess(
+    factory: ASTNodeFactory,
+    base: Expression,
+    type: StructDefinition,
+    name: string
+): MemberAccess {
+    const field = single(type.vMembers.filter((x) => x.name === name));
+    return factory.makeMemberAccess("<missing>", base, name, field.id);
+}
+
 /**
  * Transpile the `SForAll` statement `expr`. Note that this will generate and insert a loop
  * in `ctx.container`.
  */
 function transpileForAll(expr: SForAll, ctx: TranspilingContext): Expression {
     const factory = ctx.factory;
     const isOld = (ctx.semInfo.get(expr) as SemInfo).isOld;
+    const typeEnv = ctx.typeEnv;
+    let forStmt: ForStatement;
 
-    // Add temp bindings for iterator variable and result var
     const iterVarName = ctx.getForAllIterVar(expr);
     const resultVarName = ctx.getForAllVar(expr);
 
-    ctx.addBinding(iterVarName, transpileType(expr.iteratorType, ctx.factory));
+    if (expr.container !== undefined) {
+        const containerT = typeEnv.typeOf(expr.container);
+
+        // Forall over a map
+        if (containerT instanceof PointerType && containerT.to instanceof MappingType) {
+            const internalCounter = ctx.instrCtx.nameGenerator.getFresh("i");
+            const [sVar, path] = decomposeStateVarRef(expr.container);
+
+            assert(sVar !== undefined, `Unexpected undefined state var in ${expr.container.pp()}`);
+
+            const astContainer = transpile(expr.container, ctx);
+            const lib = ctx.instrCtx.getMapInterposingLibrary(
+                sVar,
+                path.map((x) => (x instanceof SNode ? null : x))
+            );
+            assert(lib !== undefined, `Unexpected missing library for map ${sVar.name}`);
+            const struct = ctx.instrCtx.getCustomMapStruct(lib);
+            const keys = makeMemberAccess(factory, astContainer, struct, "keys");
+            const len = factory.makeMemberAccess("<missing>", keys, "length", -1);
+
+            forStmt = makeForLoop(
+                ctx,
+                internalCounter,
+                new IntType(256, false),
+                new SNumber(bigInt(1), 10),
+                len
+            );
+
+            ctx.addBinding(iterVarName, transpileType(expr.iteratorType, ctx.factory));
+            (forStmt.vBody as Block).appendChild(
+                factory.makeExpressionStatement(
+                    factory.makeAssignment(
+                        "<missing>",
+                        "=",
+                        ctx.refBinding(iterVarName),
+                        factory.makeIndexAccess(
+                            "<mising>",
+                            factory.copy(keys),
+                            ctx.refBinding(internalCounter)
+                        )
+                    )
+                )
+            );
```

Comment:
Is it a case for `addStmt()`? Or, lets keep it for the followup cleaninig pass?

Reply:
Fixed in a97c7d9.